### PR TITLE
Test for id overflow and string deduplication in `SharedValueStores`

### PR DESCRIPTION
**Pull request by josh11b**
Follow-on to #3311

### CODE REVIEW

File: toolchain/base/value_store.h [hunk 1]
```diff
@@ -81,7 +81,8 @@ class ValueStore {
  public:
   // Stores the value and returns an ID to reference it.
   auto Add(typename IdT::IndexedType value) -> IdT {
-    auto id = IdT(values_.size());
+    IdT id = IdT(values_.size());
+    CARBON_CHECK(id.index >= 0) << "Too many numeric literals";
```
Comment (jonmeow) [inline toolchain/base/value_store.h]: Please fix the error message to be generic (this is not exclusive to numeric literals). Just saying "overflow" or nothing would, I think, be fine.

File: toolchain/base/value_store.h [hunk 2]
```diff
@@ -106,6 +107,7 @@ class ValueStore<StringId> {
   auto Add(llvm::StringRef value) -> StringId {
     auto [it, inserted] = map_.insert({value, StringId(values_.size())});
     if (inserted) {
+      CARBON_CHECK(it->second.index >= 0) << "Too many unique string literals";
```
Comment (jonmeow) [inline toolchain/base/value_store.h]: As above.

**Review by jonmeow — COMMENTED** [review]
Bear in mind these are generic types not for a singular purpose.

File: toolchain/base/value_store.h (hunk 2 above)
Comment (josh11b) [reply toolchain/base/value_store.h]: Isn't this the specialization for strings? It seems valuable to have this additional information.

Comment (jonmeow) [reply toolchain/base/value_store.h]: It's for strings, not necessarily string literals. Also, std::string, not Carbon strings

Comment (josh11b) [reply toolchain/base/value_store.h]: Or is this not specific to literals?

File: toolchain/base/value_store.h (hunk 1 above)
Comment (josh11b) [reply toolchain/base/value_store.h]: Ah, I see this is used for more types in later PRs.

**Review by jonmeow — APPROVED** [review]

**Pull request merged**